### ISSUE TITLE
SPAEntryPoint in library projects

### ISSUE THREAD

**Issue by cgravill** [dotnet-websharper/core#899]
We're using WebSharper to build libraries that are used in JavaScript applications. I have this working with WebSharper (4.0.194.126) but ran into an issue with 4.1.5.205.

We have a libraries like:

LibA -> LibB -> LibC

where some tools directly use the output of LibB and others LibA.

We currently using the `<WebSharperProject>Bundle</WebSharperProject>` and `<WebSharperDeadCodeElimination>False</WebSharperDeadCodeElimination>` (very keen to try the new `BundleOnly`, but that runs into other errors which I can post separately.)

To use the Bundle setting we need a `SPAEntryPoint` (but don't really want one). I workaround that by creating a project with 

```fsharp
[<JavaScript>]
module Client =

    // We must have an entry point (to bundle JavaScript out of WebSharper),
    // but we don't need it (because we're not an actual WebSharper SPA),
    // there can be only one:
    [<SPAEntryPoint>]
    let Main () = ()
```

then make that a dependency of LibA, LibB, and LibC.

I'm now getting an error:

```
1>FSC: error : Global error 'Missing entry point. Add an SPAEntryPoint attribute to a static method without arguments.' at    at WebSharper.Compiler.Packager.packageAssembly(Info refMeta, Info current, Boolean isBundle)   at WebSharper.Compiler.FrontEnd.CreateBundleJSOutput(Info refMeta, Info current)   at System.Lazy`1.CreateValue()   at System.Lazy`1.LazyInitValue()   at WebSharper.Compiler.Bundling.render@131(Lazy`1 currentJS, Boolean sourceMap, Tuple`2[] mapFileSources, FSharpFunc`2 getSetting, Boolean concatScripts, Context htmlHeadersContext, Expression pkg, FSharpList`1 resources, HtmlTextWriter noHtmlWriter, Lazy`1 assemblyLookup, FSharpRef`1 map, FSharpRef`1 minmap, BundleMode mode, StringWriter writer)   at WebSharper.Compiler.Bundling.content@228-1.Invoke(Unit unitVar)   at System.Lazy`1.CreateValue()   at System.Lazy`1.LazyInitValue()   at WebSharper.Compiler.Content.WriteFile(String name, FSharpOption`1 encoding)   at WebSharper.Compiler.Bundling.writeMapped@261(String outputDir, String fileName, Content c, FSharpOption`1 m, String ext)   at WebSharper.Compiler.Bundling.Bundle(WsConfig config, FSharpList`1 refMetas, Info currentMeta, Lazy`1 currentJS, IEnumerable`1 sources, FSharpList`1 refAssemblies)   at WebSharper.FSharp.Program.Compile$cont@185-2(WsConfig config, Boolean isBundleOnly, Loader loader, FSharpList`1 refs, Task`1 refMeta, Compilation comp, Unit unitVar)   at WebSharper.FSharp.Program.Compile$cont@64(WsConfig config, WarnSettings warnSettings, String thisName, Unit unitVar)   at WebSharper.FSharp.Program.compileMain$cont@245(String[] argv, Unit unitVar)   at WebSharper.FSharp.Program.main(String[] argv)
```

... presumably because the entry point has to exist in LibA itself not a dependency. Is that a deliberate change?

Any suggestions of how to workaround this? When I previously tried putting `SPAEntryPoint` in every library, I got errors about not being able to have multiple ones.

I was wondering, wouldn't people with `WebSharperDeadCodeElimination`
set to false, generally also not be wanting `SPAEntryPoint`? Or do you still want to be able to trigger default initialisation?

**Comment by Jand42** [comment]
@cgravill Yes, this seems an unintended side effect of the optimization in #891: when DCE is off, it concatenates existing .js files for faster bundling. It still needs to add the current .js too, and when it writes that, it only looks at current assembly for entry point.

I think the best solution would be is to add proper support for creating single `js` output that is intended to be used as a standalone library. My proposal:

* Project type remains a Bundle/BundleOnly, but a new `JavaScriptExport` attribute allows defining the scope of the API to be exposed for the assembly. Examples:
    * `[<assembly: JavaScriptExport>]` marks all of current assembly to be always fully included in output.
    * `[<assembly: JavaScriptExport("AssemblyName")>]` marks another assembly to be always fully included in output.
    * `[<assembly: JavaScriptExport(typeof<T>)>]` marks a type to be always fully included in output.
* If there are any `JavaScriptExport` attributes then `SPAEntryPoint` is optional. But the two can also coexist: this creates .js for an SPA, with the extra exports not DCE-d away (useful for debugging helpers).

DCE would then work on a starting node set determined by all/any `SPAEntryPoint` and `JavaScriptExport` attributes. Main advantage of this over current workaround of turning off DCE: all of current API is available in output, but only the minimal needed part of WebSharper core library code is included. Also, resources (external scripts or script links) can be explored for the entire scope, not just for the entry point. (Currently external resources are still determined by dependencies of the entry point, turning off DCE only affects the .js output pruning.)

Note: WebSharper currently does not track member accessibility modifiers, so JS output has all private/public members in the scope. Access modifiers carried over to output is planned for WebSharper 5.0 (TypeScript output).

Alternative, quicker fix:

* Just make `SPAEntryPoint` optional for bundles, and use current workaround of turning off DCE.

On unrelated note: if you use libraries like `LibA -> LibB -> LibC`, then LibA and LibB can't be BundleOnly, as they are needed as proper dlls too with WebSharper metadata included.

**Comment by cgravill** [comment]
@Jand42 thanks, that sounds like a thorough solution and we'd like to use DCE so would make it better than what we had before.

**Comment by Jand42** [comment]
@cgravill I am keeping this ticket open, waiting for any feedback on how much the current implementation of `JavaScriptExport` is fitting your use case. You can close if it works well, thanks a lot!

**Comment by cgravill** [comment]
@Jand42 this is looks good for our smaller projects and I've been able to move over and re-enable dead code elimination too!

On a bigger project, I'm missing some implementations. The situation is the above setup with:
LibA -> LibB -> LibC

and I'm missing LibC implementations. The intermediate project (LibB) is set to `Bundle`. Is that also a problem as you caution above about `BundleOnly`?

(by the way we're interpreting my `->` the opposite way around, I mean it as LibA depends on LibB rather than feeds into, but it makes sense either way)

**Comment by Jand42** [comment]
@cgravill As it is implemented, `JavaScriptExport` only affects output of current project, and only the occurrences of this attribute in current project are counted. This is so that multiple cross-cutting exports can be created from the same implementation projects.

However, you can target a type/module or a whole assembly in another project with: `assembly: JavaScriptExport("LibC")` or `assembly: JavaScriptExport(typeof<LibC.MyType>)`/`assembly: JavaScriptExport("LibC.MyModule+MySubModule")`. Define these in `LibB` which has the bundle output.

Does this make sense? Maybe this is a bit counter-intuitive, and although it offers some flexibility, exporting a single method in another assembly is not yet possible without exporting the full containing type.

**Comment by cgravill** [comment]
@Jand42 I think we might be talking about two close but separate parts. I'll try to give more details and you can correct me if I've got it mixed up.

In the LibB, there are several functions which we use from TypeScript. So I mark them as `JavaScriptExport` in their implementations they make use of other functions in LibC, those functions are marked with `JavaScript` and everything seems to work without marking LibC functions with export. The implementations from LibC end up in the bundled JavaScript and therefore it works at run-time: Function-in-B (exported) then calls into Function-in-C.

When I extend this process so that LibA has functions that are exported calls, that internally call into functions in LibB, which in turn calls into functions in LibC something goes wrong. It looks like the transitive implementation dependencies are getting missed somehow.

There's a lot of code involved so something else might be happening but it is the case in project-B's .js I get functions form LibC which in project-A's .js are missing.

It sounds like I could workaround the current problem by in project-A also exporting the entry points in LibB, however, that would be a bit fragile as it may be the case that there's code in LibB that's only needed by project-a not project-b and its api.

**Comment by cgravill** [comment]
I can try to work up a minimal reproduction if that would be helpful?

**Comment by Jand42** [comment]
@cgravill Sorry for the slow reply, I have been unable to reproduce your error yet. I have created a minimal testing project for this: https://github.com/Jand42/multi-bundle-test

* A = Library1
* B = SPA1
* C = SPA2

The `UsedIndirectlyIn2` type is not showing up in `SPA1.js` but does in `SPA2.js` as expected.

Have your projects had runtime failures, because only functions from LibC were missing while the LibA ones that would need it indirectly (and the LibB ones using it directly) was exported correctly?

**Comment by cgravill** [comment]
Yes, @Jand42 I've had runtime failures on this. Sounds like it's something else causing the issue. Thanks for trying to reproduce it.

I'll try to narrow down the behaviour.

**Comment by cgravill** [comment]
Hi, sorry for the delay. I couldn't reproduce the behaviour with SPA to SPA links. It appears it's only an issue with bundle to bundle links.

![image](https://user-images.githubusercontent.com/1099725/43517604-431d1fac-9581-11e8-925f-3bfab5a506ae.png)

What I did there was create a chain of projects of type Bundle. Unlike the SPAs which have guards

```JavaScript
Library1=Global.Library1=Global.Library1||{};
```

then the bundle projects count on the global being there and end up as undefined instead, and lead to runtime failures

- A = Library1
- B = Bundle1
- C = Bundle2


Bundle1
```FSharp
[<JavaScript>]
module Bundle1 =

    [<JavaScript>]
    let wrapsIndirect() =
        let matrix = Library1.Matrix.zeros
        7
```

Bundle2
```FSharp
[<JavaScriptExport>]
type Class1() = 
    member this.X = "F#"

    member this.Test =
        Bundle1.Bundle1.wrapsIndirect()
```

**Comment by Tarmil** [comment]
I think there's a problem with your setup, your "bundle" projects seem to be actually compiled as Library. In fact, your Bundle1 shouldn't compile at all if it is indeed a bundle, because it contains neither `SPAEntryPoint` nor `JavaScriptExport`.

How exactly did you set them up? I'm guessing you created Library projects and then edited the fsproj files? If so, then this is probably getting overridden by `wsconfig.json` to still compile as a library. You need to edit `wsconfig.json` like so:

```json
{
    "project": "bundle",
    "outputDir": "Content"
}
```

And then grab the js files from the `Content` folder instead of extracting them from the compiled assembly (as I assume you're currently doing?)

**Comment by cgravill** [comment]
Correct guess on creating the projects, I then removed the relevant parts of `wsconfig.json` so they wouldn't override - I like the `wsconfig.json` approach, but I was trying to match existing projects that existed before `wsconfig.json`.

I am grabbing those files from the `Content` directory, with the above set-up I still get them out without having to manual extract from the dll.

I'll try to move it more in-line with what you're currently providing as that might clear up the issues. If it persists I'll clean it up enough to post as a test repository.

**Comment by cgravill** [comment]
I'm finding it difficult to reproduce the exact issue, even after switching to `wsconfig.json` and referencing our internal library from the test projects.

I've found a related issue which might help narrow it down.

In the above set-up if don't have a reference Bundle2 to Library1 (which it only indirectly depends on), then it will build (without warning or error) but generate incomplete code. There's a repository here that reproduces this:
https://github.com/cgravill/websharper-multi-bundle-test

![image](https://user-images.githubusercontent.com/1099725/44030983-6816f422-9efa-11e8-8820-0e7fc42c57d3.png)

It's exactly the same outward behaviour as above, so I suspect in our real code something related is happening. Are you able to detect situations like the above and explicitly error with more information?

**Comment by Tarmil** [comment]
Thanks for the repro repository, I do see the issue happening with it too.

We should probably throw an error at compile time when a reference is missing like this. The irony is that we do throw a compile-time error but only if DCE is enabled, because `trimMetadata` implicitly checks the existence of all nodes accessible from the entry point and/or exports.

@Jand42 Thoughts?

**Comment by cgravill** [comment]
Is DCE off by default for bundles? I hadn't tried to turn it off deliberately on the sample and thought it was on by default - you added the switch for us previously https://github.com/dotnet-websharper/core/issues/683#issuecomment-292996629

If you were able to throw the error on missing nodes it might make it easier tracking down what's going wrong on our much larger solution.

**Comment by Jand42** [comment]
DCE is on by default (check both your project file and wsconfig.json if you have WS settings in both). If it is off, the indeed, transitive dependencies which are not themselves referenced can cause runtime problems (same with Sitelets projects).

As @Tarmil mentioned, the tree shaking is catching these errors, so it would be possible to change WS to do that even when DCE is off to expose errors and then just ignore the main result and go on printing everything. Tradeoff is some increase in compilation time, but this check can be parallelized with output printing too.

**Comment by cgravill** [comment]
I'm not sure when this resolved but it's now working correctly with 4.5.5.319! Thanks for all your help with this.
